File: benchmark/report.go
package benchmark

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

func (r *Results) SaveReport(outputDir string) error {
	timestamp := time.Now().Format("20060102-150405")
	
	// Save JSON report
	jsonPath := filepath.Join(outputDir, fmt.Sprintf("4corners-report-%s.json", timestamp))
	jsonData, err := json.MarshalIndent(r, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal JSON: %v", err)
	}
	
	err = os.WriteFile(jsonPath, jsonData, 0644)
	if err != nil {
		return fmt.Errorf("failed to write JSON report: %v", err)
	}
	
	// Save text report
	textPath := filepath.Join(outputDir, fmt.Sprintf("4corners-report-%s.txt", timestamp))
	textReport := r.GenerateTextReport()
	err = os.WriteFile(textPath, []byte(textReport), 0644)
	if err != nil {
		return fmt.Errorf("failed to write text report: %v", err)
	}
	
	return nil
}

func (r *Results) GenerateTextReport() string {
	report := "========================================\n"
	report += "4Corners Disk Benchmark Report\n"
	report += "========================================\n\n"
	report += fmt.Sprintf("Test Date: %s\n", r.TestDate.Format("2006-01-02 15:04:05"))
	report += fmt.Sprintf("Device: %s\n\n", r.Config.Device)
	
	report += "Configuration:\n"
	report += fmt.Sprintf("  Read Throughput IO Size: %s\n", r.Config.ReadTPIOSize)
	report += fmt.Sprintf("  Write Throughput IO Size: %s\n", r.Config.WriteTPIOSize)
	report += fmt.Sprintf("  Read IOPS IO Size: %s\n", r.Config.ReadIOPSIOSize)
	report += fmt.Sprintf("  Write IOPS IO Size: %s\n\n", r.Config.WriteIOPSIOSize)
	
	report += "Results:\n"
	report += "========================================\n"
	report += "Read Throughput Test:\n"
	if r.ReadTPThreads > 0 {
		report += fmt.Sprintf("  Threads:    %10d\n", r.ReadTPThreads)
		report += fmt.Sprintf("  Duration:   %10d seconds\n", r.ReadTPDuration)
	}
	report += fmt.Sprintf("  Throughput: %10.2f MB/s\n", r.ReadThroughputMBps)
	report += fmt.Sprintf("  IOPS:       %10.0f IOPS\n", r.ReadThroughputIOPS)
	report += fmt.Sprintf("  Latency:    %10.2f ms\n\n", r.ReadTPLatencyMs)
	
	report += "Write Throughput Test:\n"
	if r.WriteTPThreads > 0 {
		report += fmt.Sprintf("  Threads:    %10d\n", r.WriteTPThreads)
		report += fmt.Sprintf("  Duration:   %10d seconds\n", r.WriteTPDuration)
	}
	report += fmt.Sprintf("  Throughput: %10.2f MB/s\n", r.WriteThroughputMBps)
	report += fmt.Sprintf("  IOPS:       %10.0f IOPS\n", r.WriteThroughputIOPS)
	report += fmt.Sprintf("  Latency:    %10.2f ms\n\n", r.WriteTPLatencyMs)
	
	report += "Read IOPS Test:\n"
	if r.ReadIOPSThreads > 0 {
		report += fmt.Sprintf("  Threads:    %10d\n", r.ReadIOPSThreads)
		report += fmt.Sprintf("  Duration:   %10d seconds\n", r.ReadIOPSDuration)
	}
	report += fmt.Sprintf("  Throughput: %10.2f MB/s\n", r.ReadIOPSThroughputMBps)
	report += fmt.Sprintf("  IOPS:       %10.0f IOPS\n", r.ReadIOPS)
	report += fmt.Sprintf("  Latency:    %10.2f ms\n\n", r.ReadIOPSLatencyMs)
	
	report += "Write IOPS Test:\n"
	if r.WriteIOPSThreads > 0 {
		report += fmt.Sprintf("  Threads:    %10d\n", r.WriteIOPSThreads)
		report += fmt.Sprintf("  Duration:   %10d seconds\n", r.WriteIOPSDuration)
	}
	report += fmt.Sprintf("  Throughput: %10.2f MB/s\n", r.WriteIOPSThroughputMBps)
	report += fmt.Sprintf("  IOPS:       %10.0f IOPS\n", r.WriteIOPS)
	report += fmt.Sprintf("  Latency:    %10.2f ms\n", r.WriteIOPSLatencyMs)
	report += "========================================\n"
	
	return report
}

File: src/report.rs
use chrono::{DateTime, Local};
use serde::Serialize;
use std::fs;
use std::io;
use std::path::Path;

#[derive(Debug, Clone, Serialize)]
pub struct TestResult {
    pub throughput_mbps: f64,
    pub iops: f64,
    pub latency_avg_us: f64,
    pub latency_p50_us: f64,
    pub latency_p99_us: f64,
    pub threads: u32,
    pub queue_depth: u32,
    pub block_size_kb: u32,
    pub duration_secs: u32,
}

#[derive(Debug, Clone, Serialize)]
pub struct BenchmarkReport {
    pub test_date: DateTime<Local>,
    pub device: String,
    pub read_throughput: Option<TestResult>,
    pub write_throughput: Option<TestResult>,
    pub read_iops: Option<TestResult>,
    pub write_iops: Option<TestResult>,
}

impl BenchmarkReport {
    pub fn new(device: &str) -> Self {
        Self {
            test_date: Local::now(),
            device: device.to_string(),
            read_throughput: None,
            write_throughput: None,
            read_iops: None,
            write_iops: None,
        }
    }

    pub fn generate_text_report(&self) -> String {
        let mut s = String::new();
        s.push_str("========================================\n");
        s.push_str("4Corners Disk Benchmark Report\n");
        s.push_str("========================================\n\n");
        s.push_str(&format!(
            "Test Date: {}\n",
            self.test_date.format("%Y-%m-%d %H:%M:%S")
        ));
        s.push_str(&format!("Device: {}\n\n", self.device));

        if let Some(r) = &self.read_throughput {
            s.push_str("Read Throughput Test:\n");
            format_result(&mut s, r);
        }
        if let Some(r) = &self.write_throughput {
            s.push_str("Write Throughput Test:\n");
            format_result(&mut s, r);
        }
        if let Some(r) = &self.read_iops {
            s.push_str("Read IOPS Test:\n");
            format_result(&mut s, r);
        }
        if let Some(r) = &self.write_iops {
            s.push_str("Write IOPS Test:\n");
            format_result(&mut s, r);
        }

        s.push_str("========================================\n");
        s
    }

    pub fn save(&self, dir: &Path) -> io::Result<()> {
        let timestamp = self.test_date.format("%Y%m%d-%H%M%S");

        let text_path = dir.join(format!("4c-report-{}.txt", timestamp));
        fs::write(&text_path, self.generate_text_report())?;
        println!("Text report saved: {}", text_path.display());

        let json_path = dir.join(format!("4c-report-{}.json", timestamp));
        let json = serde_json::to_string_pretty(self).unwrap();
        fs::write(&json_path, json)?;
        println!("JSON report saved: {}", json_path.display());

        Ok(())
    }
}

fn format_result(s: &mut String, r: &TestResult) {
    s.push_str(&format!("  Threads:         {}\n", r.threads));
    s.push_str(&format!("  Queue Depth:     {}\n", r.queue_depth));
    s.push_str(&format!("  Block Size:      {} KB\n", r.block_size_kb));
    s.push_str(&format!("  Duration:        {} seconds\n", r.duration_secs));
    s.push_str(&format!("  Throughput:    {:>10.2} MB/s\n", r.throughput_mbps));
    s.push_str(&format!("  IOPS:          {:>10.0}\n", r.iops));
    s.push_str(&format!(
        "  Avg Latency:   {:>10.2} us\n",
        r.latency_avg_us
    ));
    s.push_str(&format!(
        "  P50 Latency:   {:>10.2} us\n",
        r.latency_p50_us
    ));
    s.push_str(&format!(
        "  P99 Latency:   {:>10.2} us\n",
        r.latency_p99_us
    ));
    s.push('\n');
}
